src/target/riscv/riscv-013.c

```diff
@@ -3396,329 +3396,567 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-		riscv_program_csrrsi(&program, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN, GDB_REGNO_DCSR);
-	switch (size) {
-		case 1:
-			riscv_program_lbr(&program, GDB_REGNO_S0, GDB_REGNO_S0, 0);
-			break;
-		case 2:
-			riscv_program_lhr(&program, GDB_REGNO_S0, GDB_REGNO_S0, 0);
-			break;
-		case 4:
-			riscv_program_lwr(&program, GDB_REGNO_S0, GDB_REGNO_S0, 0);
-			break;
-		case 8:
-			riscv_program_ldr(&program, GDB_REGNO_S0, GDB_REGNO_S0, 0);
-			break;
-		default:
-			LOG_ERROR("Unsupported size: %d", size);
-			goto restore_mstatus;
+		if (register_read_direct(target, &counter, GDB_REGNO_A0) != ERROR_OK)
+			return ERROR_FAIL;
+		index_on_target = counter;
+	} else {
+		target_addr_t address_on_target;
+
+		if (register_read_direct(target, &address_on_target, GDB_REGNO_S0) != ERROR_OK)
+			return ERROR_FAIL;
+		index_on_target = (address_on_target - access.target_address) /
+			access.increment;
 	}
-	if (riscv_enable_virtual && has_sufficient_progbuf(target, 5) && get_field(mstatus, MSTATUS_MPRV))
-		riscv_program_csrrci(&program, GDB_REGNO_ZERO,  CSR_DCSR_MPRVEN, GDB_REGNO_DCSR);
 
-	if (riscv_program_ebreak(&program) != ERROR_OK)
-		goto restore_mstatus;
-	if (riscv_program_write(&program) != ERROR_OK)
-		goto restore_mstatus;
+	/* According to the spec, if an abstract command fails, one can't make any
+	 * assumptions about dm_data registers, so all the values in the pipeline
+	 * are clobbered now and need to be reread.
+	 */
+	const uint32_t min_index_on_target = start_index + 2;
+	if (index_on_target < min_index_on_target) {
+		LOG_TARGET_ERROR(target, "Arithmetic does not work correctly on the target");
+		return ERROR_FAIL;
+	} else if (index_on_target == min_index_on_target) {
+		LOG_TARGET_DEBUG(target, "No forward progress");
+	}
+	const uint32_t next_index = (index_on_target - 2);
+	*elements_read = next_index - start_index;
+	return read_memory_progbuf_inner_startup(target, access.target_address,
+			access.increment, next_index);
+}
 
-	/* Write address to S0, and execute buffer. */
-	if (write_abstract_arg(target, 0, address, riscv_xlen(target)) != ERROR_OK)
-		goto restore_mstatus;
-	uint32_t command = access_register_command(target, GDB_REGNO_S0,
-			riscv_xlen(target), AC_ACCESS_REGISTER_WRITE |
-			AC_ACCESS_REGISTER_TRANSFER | AC_ACCESS_REGISTER_POSTEXEC);
-	if (execute_abstract_command(target, command) != ERROR_OK)
-		goto restore_mstatus;
+static int read_memory_progbuf_inner_on_dmi_busy(struct target *target,
+		uint32_t start_index, uint32_t next_index,
+		struct memory_access_info access);
+/**
+ * This function extracts the data from the batch.
+ */
+static int read_memory_progbuf_inner_read_batch(struct target *target,
+		const struct riscv_batch *batch,
+		uint32_t start_index, uint32_t elements_to_read, uint32_t *elements_read,
+		struct memory_access_info access)
+{
+	const bool two_reads_per_element = access.element_size > 4;
+	const uint32_t reads_per_element = (two_reads_per_element ? 2 : 1);
```

Comment:
I think you can eliminate `two_reads_per_element` and just use `reads_per_element == 2` where you need it.

Reply:
This flag has quite a lot of uses in conditions in the function, so IMHO it is justified to make it a separate variable, since it makes all the conditions a bit simpler.
I will eliminate it, if you insist.

Reply:
OK. Leave it as is.